pygls/capabilities.py

```diff
@@ -14,30 +14,56 @@
-from pygls.lsp.types import (CodeLensOptions, CompletionOptions, DocumentLinkOptions,
-                             ExecuteCommandOptions, ImplementationOptions, SaveOptions,
-                             SemanticTokensOptions, SemanticTokensRegistrationOptions,
-                             SemanticTokensRequestsFull,
-                             ServerCapabilities, SignatureHelpOptions,
-                             TextDocumentSyncOptionsServerCapabilities, TypeDefinitionOptions,
-                             WorkspaceFileOperationsServerCapabilities,
-                             WorkspaceFoldersServerCapabilities, WorkspaceServerCapabilities)
+from functools import reduce
+from typing import Any
+
+from lsprotocol.types import (
+    TEXT_DOCUMENT_CODE_ACTION, TEXT_DOCUMENT_CODE_LENS,
+    TEXT_DOCUMENT_COMPLETION, TEXT_DOCUMENT_DECLARATION,
+    TEXT_DOCUMENT_DEFINITION, TEXT_DOCUMENT_DOCUMENT_COLOR,
+    TEXT_DOCUMENT_DOCUMENT_HIGHLIGHT, TEXT_DOCUMENT_DOCUMENT_LINK,
+    TEXT_DOCUMENT_DOCUMENT_SYMBOL, TEXT_DOCUMENT_FOLDING_RANGE,
+    TEXT_DOCUMENT_FORMATTING, TEXT_DOCUMENT_HOVER,
+    TEXT_DOCUMENT_IMPLEMENTATION, TEXT_DOCUMENT_ON_TYPE_FORMATTING,
+    TEXT_DOCUMENT_RANGE_FORMATTING, TEXT_DOCUMENT_REFERENCES,
+    TEXT_DOCUMENT_RENAME, TEXT_DOCUMENT_SELECTION_RANGE,
+    TEXT_DOCUMENT_SIGNATURE_HELP, TEXT_DOCUMENT_PREPARE_CALL_HIERARCHY,
+    TEXT_DOCUMENT_DID_CLOSE, TEXT_DOCUMENT_DID_OPEN,
+    TEXT_DOCUMENT_DID_SAVE, TEXT_DOCUMENT_LINKED_EDITING_RANGE,
+    TEXT_DOCUMENT_MONIKER, TEXT_DOCUMENT_SEMANTIC_TOKENS_FULL,
+    TEXT_DOCUMENT_SEMANTIC_TOKENS_FULL_DELTA, TEXT_DOCUMENT_SEMANTIC_TOKENS_RANGE,
+    TEXT_DOCUMENT_WILL_SAVE, TEXT_DOCUMENT_WILL_SAVE_WAIT_UNTIL,
+    TEXT_DOCUMENT_TYPE_DEFINITION, WORKSPACE_DID_CREATE_FILES,
+    WORKSPACE_DID_DELETE_FILES, WORKSPACE_DID_RENAME_FILES,
+    WORKSPACE_SYMBOL, WORKSPACE_WILL_CREATE_FILES,
+    WORKSPACE_WILL_DELETE_FILES, WORKSPACE_WILL_RENAME_FILES
+)
+from lsprotocol.types import (
+    ClientCapabilities, CodeLensOptions, CompletionOptions,
+    DocumentLinkOptions, ExecuteCommandOptions, ImplementationOptions,
+    SaveOptions, SemanticTokensOptions, SemanticTokensRegistrationOptions,
+    SemanticTokensOptionsFullType1, ServerCapabilities,
+    ServerCapabilitiesWorkspaceType, SignatureHelpOptions,
+    TextDocumentSyncOptions, TypeDefinitionOptions,
+    FileOperationOptions, WorkspaceFoldersServerCapabilities
+)
+
+
+def get_capability(
+    client_capabilities: ClientCapabilities, field: str, default: Any = None
+) -> Any:
+    """Check if ClientCapabilities has some nested value without raising
+    AttributeError.
+    e.g. get_capability('text_document.synchronization.will_save')
+    """
+    try:
+        value = reduce(getattr, field.split("."), client_capabilities)
+    except AttributeError:
+        return default
+
+    # If we reach the desired leaf value but it's None, return the default.
+    value = default if value is None else value
+    return
```

Comment:
you've lost the actual return value here!  ie this should be `return value`

Reply:
Good spot! You would think that's something that would've been caught by the testsuite...

Reply:
I've pushed an extra commit that fixes this and added some tests that hopefully will catch something like this again in the future